src/sink.rs

```diff
@@ -0,0 +1,392 @@
+    /// Contrary to `extend_from_within`, the copy output can overlap with
+    /// the copy source bytes.
+    /// In addition, a copy with `start` == `pos` is valid and will
+    /// fill the sink `len` zero bytes.
+    /// # Panics
+    /// Panics if `start` > `pos`.
+    fn extend_from_within_overlapping(&mut self, start: usize, len: usize);
+}
+
+/// A Sink baked by a &[u8]
+pub struct SliceSink<'a> {
+    /// The working slice, which may contain uninitialized bytes
+    output: &'a mut [u8],
+    /// Number of bytes in start of `output` guaranteed to be initialized
+    pos: usize,
+}
+
+impl<'a> SliceSink<'a> {
+    /// Creates a `Sink` backed by the given byte slice.
+    /// `pos` defines the initial output position in the Sink.
+    /// # Panics
+    /// Panics if `pos` is out of bounds.
+    #[inline]
+    pub fn new(output: &'a mut [u8], pos: usize) -> Self {
+        // SAFETY: Caller guarantees that all elements of `output` are initialized.
+        let _ = &mut output[..pos]; // bounds check pos
+        SliceSink { output, pos }
+    }
+}
+
+impl<'a> Sink for SliceSink<'a> {
+    #[cfg(not(all(feature = "safe-encode", feature = "safe-decode")))]
+    unsafe fn base_mut_ptr(&mut self) -> *mut u8 {
+        self.output.as_mut_ptr()
+    }
+
+    #[inline]
+    fn filled_slice(&self) -> &[u8] {
+        &self.output[..self.pos]
+    }
+
+    #[inline]
+    fn pos(&self) -> usize {
+        self.pos
+    }
+
+    #[inline]
+    fn capacity(&self) -> usize {
+        self.output.len()
+    }
+
+    #[cfg(not(all(feature = "safe-encode", feature = "safe-decode")))]
+    #[inline]
+    unsafe fn set_pos(&mut self, new_pos: usize) {
+        debug_assert!(new_pos <= self.capacity());
+        self.pos = new_pos;
+    }
+
+    #[inline]
+    fn extend_with_fill(&mut self, byte: u8, len: usize) {
+        self.output[self.pos..self.pos + len].fill(byte);
+        self.pos += len;
+    }
+
+    #[inline]
+    fn extend_from_slice_wild(&mut self, data: &[u8], copy_len: usize) {
+        assert!(copy_len <= data.len());
+        self.output[self.pos..self.pos + data.len()].copy_from_slice(data);
+        self.pos += copy_len;
+    }
+
+    #[inline]
+    fn extend_from_within_wild(&mut self, start: usize, wild_len: usize, copy_len: usize) {
+        // Safety checks so that set_pos later don't expose uninitialized data
+        assert!(copy_len <= wild_len);
+        assert!(start + copy_len <= self.pos);
+        self.output.copy_within(start..start + wild_len, self.pos);
+        self.pos += copy_len;
+    }
+
+    #[inline]
+    fn extend_from_within_overlapping(&mut self, start: usize, len: usize) {
+        // Sink safety invariant guarantees that the first `pos` items are always initialized.
+        assert!(start <= self.pos);
+        let offset = self.pos - start;
+        let out = &mut self.output[start..self.pos + len];
+        out[offset] = 0; // ensures that a copy w/ start == pos becomes a zero fill
```

Comment:
why do we need this?

Reply:
It's to prevent Sink from exposing data that was previously in the buffer when `start == pos` (`offset == 0`).

_Strictly_ speaking it isn't "rust unsafe" since it's an implementation over an "at some point initialized" &[u8] BUT it could still expose previous buffer data as "decompressed" data, which is still arguably unsafe.

Do you think it should it be removed?

Reply:
can't we assert start != pos ?

Reply:
Yes, we could fail decompression with an error. A panic doesn't pass fuzz tests and that's hoe I found this problem. 

At some point I had a extra branch (offset = 0) that failed with InvalidOffset but that regressed a couple percentage points.

Then I checked the reference implementation and noticed that they do the zero fill strategy, so I copied the technique as it was also faster.

Reply:
Okay, so this is for the unsafe + checked-decode case.

With the assert we already have a branch here. I think we could change the assert to return an error, could be the same performance wise.

Reply:
I don't understand. Why not behave the same as the reference implementation? And the same as the unchecked decoder?

Reply:
To recap, I think the zero-fill behavior is better because: it results in the same behavior across all features/sinks and also the reference implementation.
That being said, this PR is maybe dragging for too long so just let me know what you'd like and I'll change it.

These benchmark runs are w/ processor turbo boost disabled and picking the best throughput out of 3 runs, otherwise it's impossible to see through the noise.

Current
BlockDecompress/lz4_flex_rust/725 2.2220 GiB/s
BlockDecompress/lz4_flex_rust/34308 1.6737 GiB/s

W/o zero fill, error on offset == 0
BlockDecompress/lz4_flex_rust/725 2.0983 GiB/s
BlockDecompress/lz4_flex_rust/34308 1.6951 GiB/s

Reply:
I thought it's better to return an error on corrupted data instead of zero filling, but I let you decide on that